pkg/model/api/v1/role.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2021 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+
+	modelAPI "github.com/perses/perses/pkg/model/api"
+)
+
+type RoleInterface interface {
+	GetMetadata() modelAPI.Metadata
+	GetRoleSpec() RoleSpec
+}
+
+type ActionKind string
+
+const (
+	KindWildcard ActionKind = "*"
+	KindRead     ActionKind = "read"
+	KindCreate   ActionKind = "create"
+	KindUpdate   ActionKind = "update"
+	KindDelete   ActionKind = "delete"
+)
+
+type Permission struct {
+	// Action of the permission (read, create, update, delete, ...)
+	Action ActionKind `json:"action" yaml:"action"`
```

Comment:
probably we want a list here so we can have a role that cover multiple actions, like `create`, `read`

Reply:
Roles have a list of permissions. One permission per action kind, then any scope you want. 
It will be easier to manage when you want to add a permission to a scope.
ie: add read perm to var

You check if the role has already a perm with read action and add variable to scope list, else create new read perm with var in scope list

If permissions have list of actions, I don't know how you will select which permission to add scope.

Really easier with only one action

Reply:
> If permissions have list of actions, I don't know how you will select which permission to add scope.

if you have 

```yaml
permissions:
- actions:
  - read
  - create
  scopes:
  - "Dashboards"
  - "Variables"
```

it just means you can `create` and `read` `Dashboards` and `Variables` I don't see why it's more complexe. And it gives more flexibility from the user perspective.

Besides, since you have an array of `Permission`, that's already possible, it's just longer to do it.

For the moment to be able to do what I did above, I need to do that:

```yaml
permissions:
- actions:
  - read
  scopes:
  - "Dashboards"
  - "Variables"
- actions:
  - create
  scopes:
  - "Dashboards"
  - "Variables"
```

At the end that's the same thing, it's just longer to write. So from my perspective, you are able to manage what is currently proposed in the PR, you should be able to manage what I am suggesting (aka an array of action)

Reply:
Totally agree it's easier for users, but I think it will be more tricky to implem. Let's see

Reply:
tricky doesn't mean impossible, it's just mean you need to think a bit more ;)